lib/Language/Haskell/Stylish/Step/Imports.hs

```diff
@@ -160,6 +201,63 @@ formatImports maxCols options moduleStats rawGroup =
   forM_ group \imp -> printQualified options padNames stats imp >> newline
 
 
+--------------------------------------------------------------------------------
+groupAndFormat
+  :: Maybe Int
+  -> Options
+  -> ImportStats
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+  -> Editor.Edits
+groupAndFormat maxCols options moduleStats groups = edits
+  where edits = Editor.changeLines block (const regroupedLines)
+        regroupedLines =
+          intercalate [""] $
+          map (formatImports maxCols options moduleStats) grouped
+
+        grouped = groupByPatterns (groupPatterns options) imports
+        imports = concatMap toList groups
+
+        block = Block
+          (GHC.srcSpanStartLine . src $ head imports)
+          (GHC.srcSpanEndLine   . src $ last imports)
+        src = fromMaybe (error "regroupImports: missing location") .
+          GHC.srcSpanToRealSrcSpan . GHC.getLocA
+
+groupByPatterns
+  :: [Pattern]
+  -> [GHC.LImportDecl GHC.GhcPs]
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+groupByPatterns patterns allImports = go patterns allImports []
+  where
+    go [] [] groups            = groups
+    go [] imports groups       = groups <> [NonEmpty.fromList imports]
```

Comment:
Almost certainly doesn't matter given the sizes of lists we're looking at, but this function does a lot of concatenating things on the back of lists, which is linear. I'd either use a `DList` or build it in reverse order and reverse at the end.

Reply:
Great call. Avoiding quadratic things seems like the way to go even if it doesn't matter now—doesn't take *that much* to turn it into a real problem.